firmware/Components.h

```diff
@@ -68,92 +71,94 @@ namespace platypus
 
-    bool set(const char* param, const char* value);
-    virtual char *name() = 0;
-    
-    void scale(float scale);
-    float scale();
-    
-    void offset(float offset);
-    float offset();
-    
   private:
-    float scale_;
-    float offset_;
+    const int measurementInterval;
+    int lastMeasurementTime;
+    double lastReading;
+  };
+
+  class IMU : public Sensor
+  {
+  public:
+    IMU(int id);
+    virtual char* name();
+    void loop();
+
+  private:
+    const int measurementInterval;
+    int lastMeasurementTime;
+    Adafruit_BNO055 bno;
+
+    uint8_t sysCalib;
+    uint8_t gyroCalib;
+    uint8_t accelCalib;
+    uint8_t magCalib;
   };
   
-  class ServoSensor : public Sensor 
+  // External Sensors //
+  class EmptySensor :  public ExternalSensor
+  {
+  // Empty Sensor placeholder representing a non-connected port
+  public:
+    EmptySensor(int id) : EmptySensor(id, id){};
+    EmptySensor(int id, int port) : ExternalSensor(id, port){};
+
+    virtual char *name(){ return "empty"; };
+  };
+
+  class ServoSensor : public ExternalSensor 
   {
   public:
-    ServoSensor(int channel);
+    ServoSensor(int id) : ServoSensor(id, id){};
+    ServoSensor(int id, int port);
     ~ServoSensor();
 
     bool set(const char* param, const char* value);
     virtual char *name();
     
     void position(float velocity);
-    float position();
+    float position(){ return position_; };
     
   private:
     Servo servo_;
     float position_;
   };
 
-  class PoweredSensor : virtual public Sensor 
-  {
-  public:
-    PoweredSensor(int channel, bool poweredOn=true);
-    virtual char *name() = 0;
-    bool powerOn();
-    bool powerOff();
-
-  private:
-    bool state_;
-  };
-
-  class SerialSensor : virtual public Sensor
+  class ES2 : public PoweredSensor, public SerialSensor
   {
   public:
-    SerialSensor(int channel, int baudRate, int serialType = RS232, int dataStringLength = 0);
-    virtual char * name() = 0;
-    void onSerial();
-
-    enum SERIAL_TYPE{
-      RS232,
-      RS485
-    };
-
-  protected:
-    int baud_;
-    int serialType_;
-    int minDataStringLength_;
-    char recv_buffer_[DEFAULT_BUFFER_SIZE];
-    unsigned int recv_index_;
-  };
+    ES2(int id) : ES2(id, id){};
+    ES2(int id, int port);
+    virtual char *name();
+    void loop();
 
-  
-  class ES2 : public PoweredSensor, public SerialSensor
-  {
   private:
     SensorState state;
     int lastMeasurementTime;
     const int measurementInterval;
     const int minReadTime;
-    
-  public:
-    ES2(int channel);
-    virtual char *name();
-    void loop();
   };
   
   class AtlasPH : public SerialSensor
   {
+  public:
+    AtlasPH(int id) : AtlasPH(id, id){};
+    AtlasPH(int id, int port);
+    bool set(const char * param, const char * value);
+    virtual char * name();
+    void setTemp(double temp);
+    void calibrate(int flag);
+    void loop();
+    void onSerial();
+
   private:
     const int measurementInterval;
```

Comment:
With `measurementInterval` appearing in multiple places, we may want to consider a mixin or composition to centralize the functionality in the future.